<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        let p1 = new Promise((resolve, reject) => {
            reject('wrong');
        });

        p1.then((result) => {
            console.log('succ',result);
            return 'THEN';
        }).catch((err) => {
            console.log('error',err);
            return 'CATCH';
        }).then((result) => {
            console.log('after catch',result);
        });
        
        // let p2 = new Promise((resolve, reject) => {
        //     resolve('correct');
        // });

        // p2.then((result) => {
        //     console.log('succ',result);
        //     return 'THEN';
        // }).catch((err) => {
        //     console.log('error',err);
        //     return 'CATCH';
        // }).then((result) => {
        //     console.log('after catch',result);
        // });
    </script>
    <script>
        // let res1 = 'result 01';
        // let res2 = new Promise((resolve, reject) => {
        //     resolve('result 02');
        // });
        // let res3 = {
        //     then:function(cbk) {
        //         console.log('Manually then');
        //         cbk();
        //     }
        // };

        // let p1 = Promise.resolve(res1);
        // console.log('p1', p1);
        // p1.then(res => {
        //     console.log('res1', res);
        // });

        // let p2 = Promise.resolve(res2);
        // console.log('p2', p2);
        // p2.then(res => {
        //     console.log('res2', res);
        // });

        // let p3 = Promise.resolve(res3);
        // console.log('p3', p3);
        // p3.then(res => {
        //     console.log('res3', res);
        // });
    </script>
    <script>
        // function p() {
        //     return new Promise((resolve, reject) => {
        //         // resolve('yes');
        //         reject('no');
        //     });
        // }
        // p().then((result) => {
        //     console.log(result);
        // }).catch((err) => {
        //     console.log(err);
        // }).finally(() => {
        //     console.log('finally');
        //     return Promise.resolve('BOX1');
        // }).then((res) => {
        //     console.log('after finally', res);// res 为 undefined
        //     return Promise.resolve('BOX2');
        // }).then((res) => {
        //     console.log('after then', res); // res 为 "BOX2"
        // });
    </script>
    <script>
        // function wait(timeout) {
        //     return new Promise(resolve => {
        //         setTimeout(resolve, timeout);
        //     });
        // }

        // function stt(timeout, cbk) {
        //     setTimeout(() => {
        //         cbk();
        //     }, timeout);
        // }

        // wait(3000).then(() => {
        //     console.log(1);
        //     wait(2000).then(() => {
        //         console.log(2);
        //     });
        // });

        // (async function () {
        //     await wait(3000);
        //     console.log(1);
        //     await wait(2000);
        //     console.log(2);
        // })();
    </script>
    <script>
        // function p1(char) {
        //     return new Promise((resolve, reject) => {
        //         console.log('run p1');
        //         resolve(char);
        //     });
        // }

        // function p2(char) {
        //     // return 'b';
        //     return new Promise((resolve, reject) => {
        //         console.log('run p2');
        //         reject(char + 'b');
        //     });
        // }

        // function p3(char) {
        //     return new Promise((resolve, reject) => {
        //         console.log('run p3');
        //         resolve(char + 'c');
        //     });
        // }

        // (async function() {
        //     let char1 = await p1('a');
        //     let char2 = await p2(char1);
        //     let res = await p3(char2);
        //     console.log(res);
        // })();

        // (async function () {
        //     let res = 'a';
        //     try {
        //         for (const p of [p1, p2, p3]) {
        //             res = await p(res)
        //         }
        //         console.log(res);
        //     } catch (error) {
        //         console.log('error', error);
        //     }

        // })();
    </script>
    <script>
        // function p1() {
        //     return new Promise((resolve, reject) => {
        //         console.log('run p1');
        //         reject('a');
        //     });
        // }

        // function p2() {
        //     // return 'b';
        //     return new Promise((resolve, reject) => {
        //         console.log('run p2');
        //         resolve('b');
        //     });
        // }

        // function p3() {
        //     return new Promise((resolve, reject) => {
        //         console.log('run p3');
        //         reject('c');
        //     });
        // }

        // Promise.all([p1(), p2(), p3()]).then((arr) => {
        //     console.log('success', arr); // 若三个都 resolve，拿到结果数组。
        // }).catch((err) => {
        //     console.log('error', err); // 捕获第一个出现的 reject。
        // });

        // Promise.allSettled([p1(), p2(), p3()]).then((arr) => {
        //     console.log('success', arr); // 结果数组
        // }).catch((err) => {
        //     console.log('error', err); // 永远不会执行
        // });

        // Promise.any([p1(), p2(), p3()]).then((firstRes) => {
        //     console.log('success', firstRes); // 第一个成功的数据
        // }).catch((err) => {
        //     console.log('error', err); // 'AggregateError: No Promise in Promise.any was resolved'
        // });

        // Promise.race([p1(), p2(), p3()]).then((firstRes) => {
        //     console.log('success', firstRes); 
        // }).catch((firstErr) => {
        //     console.log('error', firstErr);
        // });

        // let res = [p1, p2, p3].reduce((a, b) => {
        //     return a.then(b);
        // }, Promise.resolve());
        // res.then(v => {
        //     console.log(v);
        // }).catch(e=>{
        //     console.log(e);
        // });
    </script>
    <script>
        // function p1(v) {
        //     return new Promise(resolve => resolve(v * 2))
        // }

        // function p2(v1, v) {
        //     return new Promise(resolve => resolve(v * 2 + v1))
        // }

        // function p3(v2, v) {
        //     return new Promise(resolve => resolve(v * 2 + v2))
        // }

        // p1(1).then(v1 => {
        //     p2(v1, 2).then((v2) => {
        //         p3(v2, 3).then((v) => {
        //             console.log('final: ', v);
        //         }).catch((err) => {

        //         });
        //     }).catch((err) => {

        //     });
        // }).catch((err) => {

        // });

        // async function asc() {
        //     let v1 = await p1(1);
        //     let v2 = await p2(v1, 2);
        //     let v = await p2(v2, 3);
        //     console.log('final: ', v);
        // }
        // asc();
    </script>
    <script>
        // function alm(text) {
        //     // throw new Error('DOWN');
        //     console.log(text);
        // }
        // // setTimeout(() => {
        // //     stt('STT');
        // // }, 3000);
        // function stt(time) {
        //     return new Promise((resolve, reject) => {
        //         setTimeout(resolve, time);
        //     });
        // }
        // stt(2000).then(res => {
        //     alm('BONG');
        //     console.log('res', res);
        // }).catch(error => {
        //     console.log('err', error);
        // })
    </script>
    <script>
        // new Promise((resolve, reject) => {
        //         console.log('Initial');

        //         resolve();
        //     })
        //     .then(() => {
        //         throw new Error('Something failed');

        //         console.log('Do this');
        //     })
        //     .catch(() => {
        //         console.error('Do that');
        //     })
        //     .then(() => {
        //         console.log('Do this, no matter what happened before');
        //     });
    </script>
    <script>
        // let p1 = new Promise((resolve, reject) => {
        //     // resolve('p1 fulfilled');
        //     reject('p1 rejected');
        // });
        // let p2 = p1.then(res => {
        //     console.log('p1 res: ', res);
        //     return 'fulfilled to p2';
        // }, error => {
        //     console.log('p1 error: ', error);
        //     return 'rejected to p2';
        //     // Promise.resolve('rejected to p2');
        // }).catch(cth => {
        //     console.log('p1 catch: ', cth);
        //     return 'catch to p2';
        // });

        // let p3 = p2.then(res => {
        //     console.log('p2 res: ', res);
        //     return 'fulfilled to p3';
        // }, error => {
        //     console.log('p2 error: ', error);
        //     return 'rejected to p3';
        // });
    </script>
</body>

</html>